File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode next;
 *   ListNode(int x) {
 *     val = x;
 *     next = null;
 *   }
 * }
 */
public class Solution {
  public ListNode detectCycle(ListNode head) {
    Set<ListNode> nodeSet = new HashSet<ListNode>();
    ListNode curr = head;
    while (curr != null) {
      if (nodeSet.contains(curr)) return curr;
      nodeSet.add(curr);
      curr = curr.next;
    }
    return null;
  }
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *   Val int
 *   Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
  nodeSet, curr := make(map[*ListNode]bool), head
  for curr != nil {
    if _, ok := nodeSet[curr]; ok {
      return curr
    }
    nodeSet[curr] = true
    curr = (*curr).Next
  }
  return nil
}